blocks/handler.go

```diff
@@ -65,6 +65,11 @@ func (h *Handler) HandleSyncedBlock(ctx context.Context, peer p2p.Peer, data []b
 	}
 	// set the block ID when received
 	b.Initialize()
+
+	if b.ID().AsHash32() != expHash {
+		return fmt.Errorf("fetched wrong block hash. want %s, got %s", expHash.ShortString(), b.ID().String())
```

Comment:
it would be also optimal to drop such peer, there should be a special error for that (pubsub.ErrValidationReject)